api/pyproject.toml

```diff
@@ -0,0 +1,65 @@
+[project]
+name = "api"
+version = "0.0.0" # redundant for application
+description = "API for the Openverse project"
+readme = "README.md"
+license = {text = "MIT"}
+authors = [
+  {name = "Openverse", email = "openverse@wordpress.org" },
+]
+
+requires-python = "==3.11.*"
+dependencies = [
+  "adrf >=0.1.4, <0.2",
```

Comment:
I'm surprised that we're unable to use the `~=` notation for these versions 😮 I had assumed we'd be able to!

Also, is the fact that these aren't exact problematic? I know we'll have the lock file, is that what Renovate/dependabot will update directly?

Reply:
Renovate and Dependabot usually update lockfiles when a dependency is updated. They might/might not adjust the lower bound of the values in the `pyproject.toml` as well (they sometimes updated `Pipfile`, sometimes not), I do not know.

`~=` is supported (afaik PDM supports [all specifiers from PEP 440](https://peps.python.org/pep-0440/#version-specifiers)) I felt that the behaviour of `~=` is confusing because it changes depending on how many parts are specified in the version.

Consider `adrf`. Here `~=0.1.4` would be equivalent to `>=0.1.4,==0.1.*` which is what we want (v0 means that there could be breaking changes even on minor versions) so all good there. Then consider `aiohttp`. Here `~=3.9.3` would be equivalent to `>=3.9.3,==3.9.*` which is not what we want (for post v1, we want minor updates too), so that would need us to write `~=3.9` which then means we lose the lower bound as that is equivalent to `>=3.9,==3.*`.

Also personally `>=,<` is very easy to understand as it corresponds to my understanding of left-inclusive, right-exclusive ranges.

Again no strong thoughts, happy to use `~=` if favoured by the team.

Reply:
Ahh, that makes a lot of sense. Thanks for the explanation, I'm definitely okay with the more explicit approach you've provided here 🙂

Reply:
Agreed `>=` is a lot clearer. `~=` always struck me as a bit weird and annoying to have to think so hard about every time I read it.

Most projects I've seen using pyproject use `>=` or `==`.